### ISSUE TITLE
xgboost( ROC Curve)

### ISSUE THREAD

**Issue by refree** [refree/xgboost#2]
#xgboost 연습 _ target변수 형식은 오직 intiger만 모형에서 돌아감
#팩터, 뉴메릭 안돌아감.

setwd("C:/Users/Park Jung/Desktop/SELS/data/query")
library(xgboost)
library(Matrix)
library(DiagrammeR)
library(dplyr)
 
  df <- read.csv("7grade_join_1.csv" )
  df <-as.data.frame(df)
  for(i in 1:length(df)) {   df[,i] <- as.numeric(df[,i]) }
 

  df <- df %>%
    mutate(do_violence_in_school = ifelse(do_violence_in_school >= 0.03965, 1, 0))
  df$do_violence_in_school <- as.integer(df$do_violence_in_school)

#-------------------------------------
library(data.table)
#df <- read.csv("df.csv")
#names(df)
#summary(df)
#summary(df$do_violence_in_school)
str(df)
df$gender = as.factor(df$gender)
df$residence_code = as.factor(df$residence_code)
#df$do_violence_in_school <- as.factor(df$do_violence_in_school)
library(caret)
trainIndex   <- createDataPartition(df$do_violence_in_school, p=0.75, list=FALSE, times=1)
train        <- df[ trainIndex, ]
test         <- df[-trainIndex, ]
str(df)
Train.X  <- sparse.model.matrix(do_violence_in_school ~ .-1, data=train)
Test.X   <- sparse.model.matrix(do_violence_in_school ~ .-1, data=test)

Train.Y <- train$do_violence_in_school
Test.Y <- test$do_violence_in_school

param <- list(objective   = "binary:logistic",
              eval_metric = "error",
              max_depth   = 7,
              eta         = 0.1,
              gammma      = 1,
              colsample_bytree = 0.5,
              min_child_weight = 1)

set.seed(1234)

#Pass in our hyperparameteres and train the model 
system.time(xgb <- xgboost(params  = param,
                           data    = Train.X,
                           label   = Train.Y, 
                           nrounds = 500,
                           print_every_n = 100,
                           verbose = 1))

xgb
#https://stackoverflow.com/questions/51607664/r-script-xgboost-for-binary-classification-how-to-get-predicted-label
#_혼동행렬 도출(7시간 만에 찾음)___________________________________________________________________________________
xgb_pred <- predict(xgb, Test.X)
xgb_pred
xgb_pred_class <- as.numeric(xgb_pred > 0.5)
confusionMatrix(as.factor(xgb_pred_class), as.factor(Test.Y),positive="1")

#변수중요도 그래프_______________________________________________________________________________________

#Determine Which Features Are Most Important

#Get the trained model
model <- xgb.dump(xgb, with_stats=TRUE)

#Get the feature real names
names <- dimnames(Train.X)[[2]]

#Compute feature importance matrix
importance_matrix <- xgb.importance(names, model=xgb)[0:20] # View top 20 most important features

#Plot
xgb.plot.importance(importance_matrix)

#ROC커브 그래프________________________________________________________________
library(ROCR)

#roc커브 내방식
xgb_pred7 <- predict(xgb, Test.X, type = "prob")
xgb.pred7 <- prediction(xgb_pred7, Test.Y)
par(mfrow=c(1,1))

plot.roc.cuvexg7 <- function(xgb.pred7, title.text){
  xgb.perf7 <- performance(xgb.pred7, "tpr", "fpr")
  plot(xgb.perf7, col="green", lty=1, lwd=2,
       main=title.text, cex.main=0.6, cex.lab=0.8,xaxs="i", yaxs="i")
  abline(0,1, col="red")
         auc7 <- performance(xgb.pred7, "auc")
         auc7 <- unlist(slot(auc7, "y.values"))
         auc7 <- round(auc7,2)
         legend(0.4,0.4,legend=c(paste0("7Grade AUC: ",auc7)),cex=1,bty = "n",box.col = "white")
}

plot.roc.cuvexg7(xgb.pred7, title.text = "RF ROC Curve");par(new=TRUE)

## Use ROCR package to plot ROC Curve // 블로그 http://jamesmarquezportfolio.com/get_up_and_running_with_xgboost_in_r.html
#xgb.pred <- prediction(xgb_pred, Test.Y)
#xgb.perf <- performance(xgb.pred, "tpr", "fpr")
# 
#plot(xgb.perf, avg="threshold",colorize=TRUE,lwd=1,main="ROC Curve w/ Thresholds", print.cutoffs.at=seq(0, 1, by=0.05),text.adj=c(-0.5, 0.5),text.cex=0.5)
#grid(col="lightgray") axis(1, at=seq(0, 1, by=0.1)) axis(2, at=seq(0, 1, by=0.1)) abline(v=c(0.1, 0.3, 0.5, 0.7, 0.9), col="lightgray", lty="dotted") abline(h=c(0.1, 0.3, 0.5, 0.7, 0.9), col="lightgray", lty="dotted")
#lines(x=c(0, 1), y=c(0, 1), col="black", lty="dotted")
